crates/crypto_api/src/crypto_system/crypto_system_test.rs

```diff
@@ -446,5 +656,222 @@ mod test {
+            }
+
+            if server_tx.len() != self.kx_session_key_bytes() {
+                return Err(CryptoError::BadTxSessionKeySize);
+            }
+
+            if server_pk.len() != self.kx_public_key_bytes() {
+                return Err(CryptoError::BadPublicKeySize);
+            }
+
+            if server_sk.len() != self.kx_secret_key_bytes() {
+                return Err(CryptoError::BadSecretKeySize);
+            }
+
+            if client_pk.len() != self.kx_public_key_bytes() {
+                return Err(CryptoError::BadPublicKeySize);
+            }
+
+            server_rx.write(0, &server_pk.read_lock()[..4])?;
+            server_rx.write(4, &client_pk.read_lock()[..4])?;
+            server_tx.write(0, &client_pk.read_lock()[..4])?;
+            server_tx.write(4, &server_pk.read_lock()[..4])?;
+
+            Ok(())
+        }
+
+        fn aead_nonce_bytes(&self) -> usize {
+            8
+        }
+
+        fn aead_auth_bytes(&self) -> usize {
+            8
+        }
+
+        fn aead_secret_bytes(&self) -> usize {
+            8
+        }
+
+        fn aead_encrypt(
+            &self,
+            cipher: &mut Box<dyn Buffer>,
+            message: &Box<dyn Buffer>,
+            adata: Option<&Box<dyn Buffer>>,
+            nonce: &Box<dyn Buffer>,
+            secret: &Box<dyn Buffer>,
+        ) -> CryptoResult<()> {
+            if cipher.len() != message.len() + self.aead_auth_bytes() {
+                return Err(CryptoError::BadCipherSize);
+            }
+
+            if nonce.len() != self.aead_nonce_bytes() {
+                return Err(CryptoError::BadNonceSize);
+            }
+
+            if secret.len() != self.aead_secret_bytes() {
+                return Err(CryptoError::BadSecretKeySize);
+            }
+
+            cipher.zero();
+            cipher.write(2, &nonce.read_lock()[..2])?;
+            cipher.write(4, &secret.read_lock()[..2])?;
+
+            if let Some(adata) = adata {
+                cipher.write(6, &adata.read_lock()[..2])?;
+            }
+
+            cipher.write(8, &message.read_lock())?;
+
+            Ok(())
+        }
+
+        fn aead_decrypt(
+            &self,
+            message: &mut Box<dyn Buffer>,
+            cipher: &Box<dyn Buffer>,
+            adata: Option<&Box<dyn Buffer>>,
+            nonce: &Box<dyn Buffer>,
+            secret: &Box<dyn Buffer>,
+        ) -> CryptoResult<()> {
+            if message.len() != cipher.len() - self.aead_auth_bytes() {
+                return Err(CryptoError::BadMessageSize);
+            }
+
+            if nonce.len() != self.aead_nonce_bytes() {
+                return Err(CryptoError::BadNonceSize);
+            }
+
+            if secret.len() != self.aead_secret_bytes() {
+                return Err(CryptoError::BadSecretKeySize);
+            }
+
+            let cipher = cipher.read_lock();
+
+            if &cipher[2..4] != &nonce.read_lock()[..2] {
```

Comment:
This code is cryptic, maybe deserves a comment about what is going here with the seemingly arbitrary array indexing.

Reply:
:+1: